paasta_tools/mesos/slave.py

```diff
@@ -36,6 +36,9 @@ def __getitem__(self, name):
     def __str__(self):
         return self.key()
 
+    def __hash__(self):
```

Comment:
This is a noop for `@async_ttl_cache(..., cleanup_self=True)`, as we don't use hash for `self` in that case.

Reply:
We just discussed IRL and I think @vkhromov is right. It might be safe for me to disable `cleanup_self` for the slaves but I want to do a bit of testing to make sure...

Reply:
As for `@async_ttl_cache(cleanup_self=False)`, it's not enough to add `__hash__()`, you also need to add `__eq__()` method to `MesosSlave`, otherwise there will be a memory leak again:
```
$ cat 3.py
#! /usr/bin/env python3.6

import functools

class WithoutEq(object):
    def __init__(self, q):
        self.q = q
    def __hash__(self):
        return 1

class WithEq(object):
    def __init__(self, q):
        self.q = q
    def __hash__(self):
        return 2
    def __eq__(self, other):
        return True

def k(self):
    return functools._make_key((self,), {}, typed=False)

def q(cls):
    d = dict()
    a1 = cls(1)
    a2 = cls(2)
    d[k(a1)] = 1
    d[k(a2)] = 2
    print(f'{d}')

q(WithoutEq)
q(WithEq)

$ ./3.py
{[<__main__.WithoutEq object at 0x7f605db61e48>]: 1, [<__main__.WithoutEq object at 0x7f605db61e80>]: 2}
{[<__main__.WithEq object at 0x7f605db61e80>]: 2}
```

Reply:
This is all correct! But I'm going to remove the `__hash__` method because I get the performance increase without it!

I worked out why:
* We call `task_stats` which calls `self.slave()` but this calls `self.master.slave(self["slave_id"])` and that call is also cached. So we only get one `MesosSlave` in the cache each time anyway.
* This leads me to conclude I don't need to add the `__hash__` or `__eq__`, I can just bump the TTL on the statistics cache.